design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,223 @@
+The annotation will still be accompanied by the feature-gate configuration.
+
+Pros:
+* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
+* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs that each VM owner needs to be aware of.
+
+## Alternative #2 Configure both pod and Guest
+It is possible to use both the current Guest API and the old yet existing pod API, like so:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        resources:
+          requests:
+            cpu: '10'
+          limits:
+            cpu: '10'
+        cpu:
+          cores: 4
+          sockets: 1
+          threads: 2
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+```
+In order to do that the current validating webhooks regarding dedicatedCpuPlacement and IsolatedEmulatorThread will have to be loosened.
+
+Pros:
+* This option does not introduce any alpha knobs.
+
+Cons:
+* This option exposes the user to unnecessary implementation details and possible waste of physical CPUs. They have to do things manually, opening the door for misuse. Loosening the validating webhook doesn't make sense in the general case.
+* Implementation difficulty: KubeVirt will need to adjust internal assumptions that are based on the fact that VMs {req, limits}.cpu = vCPUs. Specifically, how the housekeeping cgroup is being formed, but possibly other aspects.
+
+## Scalability
+The solution proposed is “wasting” one unused CPU for the purpose of maintaining SMT alignment.
+
+## Update/Rollback Compatibility
+Since the annotation is not an API change, it is backportable.
+When a better solution is possible, Kubevirt can simply replace the annotation support with a better approach, thus not making the change relatively seamless for the client. 
+
+## Functional Testing Approach
+Will add unit tests to `renderresources_test.go` to check extra housekeeping logic, and in `templates.go` to check the virt-launcher pod spec output.
+
+# Implementation Phases
+## Phase 1: Align CPU to SMT requirements when feature gate is enabled and annotation is set on the VMI.
+* Virt-config:
+  * Add support for the new feature gate: `EmulatorThreadCompleteToEvenParity`.
+* Virt-API:
+  * Add support for the new alpha annotation: `kubevirt.io/EmulatorThreadCompleteToEvenParity:`.
+  * in case annotation is set on kubevirt CR, the feature gate is enabled and the isolateEmulatorThread is enabled:
+    * the VMI mutator webhook will copy the kubevirt CR annotation to the VMI annotations.
```

Comment:
why?

Reply:
Because it adds the cluster-wide solution.
The rest is almost the same as the per-VMI solution that is already implemented by https://github.com/kubevirt/kubevirt/pull/10593 and https://github.com/kubevirt/kubevirt/pull/10783.

Reply:
The separation to phases is not really relevant anymore - since the changes to what is already merged is planned to be in one PR.